signer/keydbstore/cachedkeystore.go

```diff
@@ -0,0 +1,76 @@
+package keydbstore
+
+import (
+	"sync"
+
+	"github.com/docker/notary/trustmanager"
+	"github.com/docker/notary/tuf/data"
+)
+
+// Note:  once trustmanager's file KeyStore has been flattened, this can be moved to trustmanager
+
+type cachedKeyStore struct {
+	trustmanager.KeyStore
+	lock       *sync.Mutex
+	cachedKeys map[string]*cachedKey
+}
+
+type cachedKey struct {
+	role string
+	key  data.PrivateKey
+}
+
+// NewCachedKeyStore returns a new trustmanager.KeyStore that includes caching
+func NewCachedKeyStore(baseStore trustmanager.KeyStore) trustmanager.KeyStore {
+	return &cachedKeyStore{
+		KeyStore:   baseStore,
+		lock:       &sync.Mutex{},
+		cachedKeys: make(map[string]*cachedKey),
+	}
+}
+
+// AddKey stores the contents of a private key. Both role and gun are ignored,
+// we always use Key IDs as name, and don't support aliases
+func (s *cachedKeyStore) AddKey(keyInfo trustmanager.KeyInfo, privKey data.PrivateKey) error {
+	if err := s.KeyStore.AddKey(keyInfo, privKey); err != nil {
+		return err
+	}
+
+	// Add the private key to our cache
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	s.cachedKeys[privKey.ID()] = &cachedKey{
+		role: keyInfo.Role,
+		key:  privKey,
+	}
+
+	return nil
+}
+
+// GetKey returns the PrivateKey given a KeyID
+func (s *cachedKeyStore) GetKey(keyID string) (data.PrivateKey, string, error) {
+	s.lock.Lock()
```

Comment:
we can move this lock down to line 62. Reading from maps is thread safe, we may just end up doing some duplicate `s.KeyStore.GetKey` calls. However, if that call is slow for some reason, it'll be more efficient to not lock this whole method